ibis/expr/analysis.py

```diff
@@ -38,61 +38,16 @@ def sub_for(expr, substitutions):
     ibis.expr.types.Expr
         An Ibis expression
     """
-    mapping = {k.op(): v for k, v in substitutions}
-    substitutor = Substitutor()
-    return substitutor.substitute(expr, mapping)
-
-
-class Substitutor:
-    def __init__(self):
-        """Initialize the Substitutor class.
-
-        Notes
-        -----
-        We need a new cache per substitution call, otherwise we leak state
-        across calls and end up incorrectly reusing other substitions' cache.
-        """
-        cache = toolz.memoize(key=lambda args, kwargs: args[0]._key)
-        self.substitute = cache(self._substitute)
-
-    def _substitute(self, expr, mapping):
-        """Substitute expressions with other expressions.
-
-        Parameters
-        ----------
-        expr : ibis.expr.types.Expr
-        mapping : Mapping[ibis.expr.operations.Node, ibis.expr.types.Expr]
-
-        Returns
-        -------
-        ibis.expr.types.Expr
-        """
-        node = expr.op()
+
+    def fn(node, mapping={k.op(): v for k, v in substitutions}):
```

Comment:
It's a bit strange, how about just using fn as a closure and set mapping before the function definition?

Reply:
Done.